### DISCUSSION TITLE
LDT compile issues

### DISCUSSION THREAD

**Discussion by sunnydays99** [NASA-LIS/LISF#1443] [Q&A] [answered]
Hello I have compile LDT before, however when doing the test case I noticed I need HDF4 and now nothing is working anymore and I have been stuck for a while. That's the error code I get:

./interp/upscaleByAveraging_input.F90:16:35: note: ‘upscalebyaveraging_input’ was previously declared here
   16 | subroutine upscaleByAveraging_input( gridDesci, gridDesco, mi, mo, n11 )
      |                                   ^
../interp/upscaleByAveraging_input.F90:16:35: note: code may be misoptimized unless ‘-fno-strict-aliasing’ is used
/usr/bin/ld: /usr/lib//libmfhdf.a(putget.o): in function `xdr_NCvbyte':
putget.c:(.text+0x314): undefined reference to `xdr_opaque'
/usr/bin/ld: putget.c:(.text+0x39b): undefined reference to `xdr_opaque'
/usr/bin/ld: putget.c:(.text+0x3ec): undefined reference to `xdr_opaque'
/usr/bin/ld: /usr/lib//libmfhdf.a(putget.o): in function `sd_xdr_NCvshort':
putget.c:(.text+0x494): undefined reference to `xdr_opaque'
/usr/bin/ld: putget.c:(.text+0x51e): undefined reference to `xdr_opaque'
/usr/bin/ld: /usr/lib//libmfhdf.a(putget.o):putget.c:(.text+0x57c): more undefined references to `xdr_opaque' follow
/usr/bin/ld: /usr/lib//libmfhdf.a(putget.o): in function `xdr_NCvdata':
putget.c:(.text+0x7b3): undefined reference to `xdr_float'
/usr/bin/ld: putget.c:(.text+0x7f3): undefined reference to `xdr_int'
/usr/bin/ld: putget.c:(.text+0x803): undefined reference to `xdr_double'
/usr/bin/ld: putget.c:(.text+0x829): undefined reference to `xdr_opaque'
/usr/bin/ld: /usr/lib//libmfhdf.a(putget.o): in function `xdr_NCv1data':
putget.c:(.text+0x681): undefined reference to `xdr_int'
/usr/bin/ld: putget.c:(.text+0x699): undefined reference to `xdr_float'
/usr/bin/ld: putget.c:(.text+0x6b1): undefined reference to `xdr_double'
/usr/bin/ld: /usr/lib//libmfhdf.a(sharray.o): in function `NCxdr_shortsb':
sharray.c:(.text+0x4ce): undefined reference to `xdr_opaque'
/usr/bin/ld: /usr/lib//libmfhdf.a(sharray.o): in function `NCxdr_shortsb.constprop.0':
sharray.c:(.text+0xa9b): undefined reference to `xdr_opaque'
/usr/bin/ld: /usr/lib//libmfhdf.a(string.o): in function `sd_xdr_NC_string':
string.c:(.text+0x5ac): undefined reference to `xdr_u_long'
/usr/bin/ld: string.c:(.text+0x5e6): undefined reference to `xdr_u_long'
/usr/bin/ld: string.c:(.text+0x5fb): undefined reference to `xdr_opaque'
/usr/bin/ld: string.c:(.text+0x66c): undefined reference to `xdr_opaque'
/usr/bin/ld: string.c:(.text+0x694): undefined reference to `xdr_u_long'
/usr/bin/ld: /usr/lib//libmfhdf.a(var.o): in function `sd_xdr_NC_var':
var.c:(.text+0xdb1): undefined reference to `xdr_int'
/usr/bin/ld: var.c:(.text+0xdcd): undefined reference to `xdr_u_int'
/usr/bin/ld: var.c:(.text+0xe06): undefined reference to `xdr_u_int'
/usr/bin/ld: /usr/lib//libmfhdf.a(array.o): in function `sd_xdr_NC_array':
array.c:(.text+0xa37): undefined reference to `xdr_int'
/usr/bin/ld: array.c:(.text+0xa4b): undefined reference to `xdr_u_int'
/usr/bin/ld: array.c:(.text+0xb23): undefined reference to `xdr_opaque'
/usr/bin/ld: array.c:(.text+0xbe3): undefined reference to `xdr_int'
/usr/bin/ld: array.c:(.text+0xc8b): undefined reference to `xdr_double'
/usr/bin/ld: array.c:(.text+0xc9b): undefined reference to `xdr_float'
/usr/bin/ld: /usr/lib//libmfhdf.a(cdf.o): in function `sd_xdr_numrecs':
cdf.c:(.text+0x3926): undefined reference to `xdr_u_long'
/usr/bin/ld: /usr/lib//libmfhdf.a(cdf.o): in function `sd_xdr_cdf':
cdf.c:(.text+0x3a42): undefined reference to `xdr_u_long'
/usr/bin/ld: /usr/lib//libmfhdf.a(cdf.o): in function `sd_xdr_NC_fill':
cdf.c:(.text+0x4343): undefined reference to `xdr_bytes'
/usr/bin/ld: cdf.c:(.text+0x43bb): undefined reference to `xdr_int'
/usr/bin/ld: cdf.c:(.text+0x43cb): undefined reference to `xdr_float'
/usr/bin/ld: cdf.c:(.text+0x43e3): undefined reference to `xdr_double'
/usr/bin/ld: /usr/lib//libmfhdf.a(cdf.o): in function `sd_xdr_numrecs':
cdf.c:(.text+0x3959): undefined reference to `xdr_u_long'
/usr/bin/ld: /usr/lib//libmfhdf.a(dim.o): in function `sd_xdr_NC_dim':
dim.c:(.text+0x7d9): undefined reference to `xdr_long'
/usr/bin/ld: /usr/lib//libmfhdf.a(iarray.o): in function `sd_xdr_NC_iarray':
iarray.c:(.text+0x145): undefined reference to `xdr_u_long'
/usr/bin/ld: iarray.c:(.text+0x16b): undefined reference to `xdr_int'
/usr/bin/ld: iarray.c:(.text+0x1ac): undefined reference to `xdr_u_long'
/usr/bin/ld: iarray.c:(.text+0x1fb): undefined reference to `xdr_int'
collect2: error: ld returned 1 exit status
make: *** [Makefile:54: LDT] Error 1

[ERR] Compile failed

any ideas how to fix it? I have used HDF-4.2.16 but the same error occurred with the previous version

Kind reads and thank you for any help!!

**Comment by jvgeiger** [comment]
Hello.  Which version of LISF are you compiling: the master branch or one of the releases?

Also, what does `ls -l /lib64/libtirpc*` report?

**Comment by sunnydays99** [comment]
Hello, I am using the public version 7.4.3 and 

ls -l /lib64/libtirpc*  yields no such file and directory, I think I must have saved libtirpc wrong then, would that explain the failure in linking?

**Reply by jvgeiger** [reply]
Yes, we must find your RPC library.  Did you install HDF4 from source, or did you install a package?

What does `cat /usr/lib/libhdf4.settings` report?

Which Linux distibution are you using?  What does `cat /etc/os-release` report?

**Reply by sunnydays99** [reply]
Okay I see, I did install HDF4 from source and cat /usr/lib/libhdf4.settings reports following:

	    SUMMARY OF THE HDF4 CONFIGURATION
	    =================================

General Information:
-------------------
		   HDF4 Version: 4.2.16-2
		  Configured on: Mon Nov  6 16:50:34 CET 2023
		  Configured by: annika@annika-virtual-machine
		 Configure mode: production
		    Host system: x86_64-unknown-linux-gnu
              Uname information: Linux annika-virtual-machine 6.2.0-35-generic #35~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Fri Oct  6 10:23:26 UTC 2 x86_64 x86_64 x86_64 GNU/Linux
		      Libraries: 
	     Installation point: /usr

Compiling Options:
------------------
                     C compiler: /usr/bin/gcc
                         CFLAGS:  -std=c99 -Wno-error=implicit-function-declaration -Wlarger-than=2560 -Wlogical-op -Wframe-larger-than=16384 -Wpacked-bitfield-compat -Wsync-nand -Wno-unsuffixed-float-constants -Wdouble-promotion -Wtrampolines -Wstack-usage=8192 -Wmaybe-uninitialized -Wno-jump-misses-init -Wstrict-overflow=2 -Wno-suggest-attribute=const -Wno-suggest-attribute=noreturn -Wno-suggest-attribute=pure -Wno-suggest-attribute=format -Wdate-time -Warray-bounds=2 -Wc99-c11-compat -Wincompatible-pointer-types -Wint-conversion -Wshadow -Wduplicated-cond -Whsa -Wnormalized -Wnull-dereference -Wunused-const-variable -Walloca -Walloc-zero -Wduplicated-branches -Wformat-overflow=2 -Wformat-truncation=1 -Wrestrict -Wattribute-alias -Wcast-align=strict -Wcast-function-type -Wmaybe-uninitialized -Wshift-overflow=2 -Wno-suggest-attribute=cold -Wno-suggest-attribute=malloc -Wattribute-alias=2 -Wmissing-profile -Wc11-c2x-compat -I/usr/include/tirpc -O3 -fomit-frame-pointer
                       CPPFLAGS: -D_POSIX_C_SOURCE=200809L  -I/usr/include/tirpc -DNDEBUG 
               Shared Libraries: no
               Static Libraries: yes
  Statically Linked Executables: no
                        LDFLAGS: 
                Extra libraries: -ljpeg -lz  -ltirpc
                       Archiver: ar
                         Ranlib: ranlib

Languages:
----------
                        Fortran: yes
               Fortran Compiler: /usr/bin/gfortran
                         FFLAGS:  -fallow-argument-mismatch -O

                           Java: no

Features:
---------
               SZIP compression: disabled
 Export HDF4-built netCDF-2 API: no (yes: export undecorated netCDF names, no: prefix with 'sd_')
    HDF4-built ncdump and ncgen: yes


The Linux distribution I am using is following:

PRETTY_NAME="Ubuntu 22.04.3 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.3 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy

**Reply by jvgeiger** [reply]
Let's try:

```
ls -l /lib/libtirpc*
ls -l /usr/lib/libtirpc*
ls -l /usr/lib64/libtirpc*
```

One of those should report something.

**Reply by sunnydays99** [reply]
I'm afraid none of these report anything. I tried to locate the libtirpc* files and they are found in the subfolder x86_64-linux-gnu
$ ls -l /lib/x86_64-linux-gnu/libtirpc*
-rw-r--r-- 1 root root 404664 jul 25  2022 /lib/x86_64-linux-gnu/libtirpc.a
lrwxrwxrwx 1 root root     39 jul 25  2022 /lib/x86_64-linux-gnu/libtirpc.so -> /lib/x86_64-linux-gnu/libtirpc.so.3.0.0
lrwxrwxrwx 1 root root     17 okt 30 08:37 /lib/x86_64-linux-gnu/libtirpc.so.3 -> libtirpc.so.3.0.0
-rw-r--r-- 1 root root 182912 jul 25  2022 /lib/x86_64-linux-gnu/libtirpc.so.3.0.0

**Reply by jvgeiger** [reply, marked as answer]
Please try this.  Edit your ldt/make/configure.ldt file.  You want to change the LDFLAGS line by adding `-ltirpc` like this:

`-L$(LIB_HDF4) -lmfhdf -ldf -ljpeg -lz -ltirpc`

Then rerun `./compile` in the ldt subdirectory.

**Reply by sunnydays99** [reply]
that has done the trick! Thank you so much, LDT successfully compiled now

**Reply by jvgeiger** [reply]
Great!  If you need HDF4 support in LIS or LVT, then you must apply the same trick.

**Reply by jvgeiger** [reply]
By the way, a better fix for this problem was merged into our master branch.  If you ever switch to our master branch and then get the same linking error, try setting the LDT_RPC environment variable to the full name of your RPC library.  In your case,

`export LDT_RPC=/lib/x86_64-linux-gnu/libtirpc.so`

Similarly for LIS_RPC and LVT_RPC.
